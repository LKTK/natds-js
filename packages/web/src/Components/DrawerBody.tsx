import React, { FunctionComponent, forwardRef, useState } from "react";
import { withTheme } from "@material-ui/core";
import styled from "styled-components";
import { IThemeWeb } from "Themes";
import { getDefaultTheme } from "./shared";

export interface IDrawerBodyProps {
  children?: React.ReactNode;
  component?: React.ElementType;
  theme: IThemeWeb | unknown;
}

export const DrawerBody: FunctionComponent<IDrawerBodyProps> = forwardRef((
  props: IDrawerBodyProps,
  ref: any
) => {
  const { children, component, "theme": providerTheme, ...rest } = props;
  const [
    scrolled,
    setScrolled
  ] = useState(false);
  const theme: any = React.useMemo(() => getDefaultTheme(providerTheme), [providerTheme]);
  const SCROLL_POSITION_ZERO = 0;
  const handleScrolled = (event: any) => setScrolled(event.target.scrollTop > SCROLL_POSITION_ZERO);

  return (
    <DrawerBodyComponent
      scrolled={scrolled}
      as={component}
      theme={theme}
      ref={ref}
      {...rest}>
      <DrawerBodyScrollComponent
        onScroll={handleScrolled}
        theme={theme}>
        {children}
      </DrawerBodyScrollComponent>
    </DrawerBodyComponent>
  );
});

export default withTheme(DrawerBody);

/**
 * @todo refactor(web): replace string by shadow token
 */
const SCROLLED_BOX_SHADOW = "rgba(0, 0, 0, .14) inset 0 9px 5px -5px";

/**
 * @todo refactor(web): replace string by shadow token
 */
const NOT_SCROLLED_BOX_SHADOW = "none";

const getDrawerBodyComponentBoxShadow = ({ scrolled }: { scrolled: boolean }) => {
  if (scrolled) {
    return SCROLLED_BOX_SHADOW;
  }

  return NOT_SCROLLED_BOX_SHADOW;

};

export const DrawerBodyComponent = styled.div<{ scrolled: boolean, theme: IThemeWeb }>`
  flex: 1 1 auto;
  display: flex;
  position: relative;
  overflow: hidden;

  &:before {
    content: "";
    box-shadow: ${({scrolled}) => getDrawerBodyComponentBoxShadow({scrolled})};
    height: 10px;
    pointer-events: none;
    top: 0;
    left: 0;
    right: 0;
    position: absolute;
    transition: box-shadow ease-in-out .1s;
    z-index: 2;
  }
`;

export const DrawerBodyScrollComponent = styled.div<{ theme: IThemeWeb }>`
  flex: 1 1 auto;
  overflow-y: auto;
  padding: ${({ theme }) => `0 ${theme.sizes.tiny}px`};
  position: relative;
`;
