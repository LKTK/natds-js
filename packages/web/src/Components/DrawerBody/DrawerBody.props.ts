import * as React from "react";

export interface IDrawerBodyProps {
  children?: React.ReactNode;
  component?: React.ElementType;
}
