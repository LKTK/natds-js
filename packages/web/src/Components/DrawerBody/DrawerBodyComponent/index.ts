import DrawerBodyComponent from "./DrawerBodyComponent";

export { DrawerBodyComponent };
export default DrawerBodyComponent;
