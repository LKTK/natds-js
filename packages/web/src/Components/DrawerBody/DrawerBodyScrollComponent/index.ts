import DrawerBodyScrollComponent from "./DrawerBodyScrollComponent";

export { DrawerBodyScrollComponent };
export default DrawerBodyScrollComponent;
