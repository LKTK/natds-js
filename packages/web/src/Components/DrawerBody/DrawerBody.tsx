import * as React from "react";
import { IDrawerBodyProps } from "./DrawerBody.props";
import { DrawerBodyComponent } from "./DrawerBodyComponent";
import { DrawerBodyScrollComponent } from "./DrawerBodyScrollComponent";
import useDefaultTheme from "../../hooks/useDefaultTheme";

export { IDrawerBodyProps } from "./DrawerBody.props";

/**
 * ## Importing
 *
 * ```
 * import { DrawerBody } from '@naturacosmeticos/natds-web';
 * ```
 */
export const DrawerBody = React.forwardRef<HTMLElement, IDrawerBodyProps>((
  props: IDrawerBodyProps, ref,
) => {
  const {
    children, component, ...otherProps
  } = props;
  const [
    scrolled, setScrolled,
  ] = React.useState(false);
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const theme: any = useDefaultTheme();
  const SCROLL_POSITION_ZERO = 0;

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const handleScrolled = (event: any) => {
    if (typeof event === "undefined") {
      return;
    }
    const { target } = event;

    if (typeof target === "undefined") {
      return;
    }
    const { scrollTop } = target;

    if (typeof scrollTop !== "undefined") {
      setScrolled(scrollTop > SCROLL_POSITION_ZERO);
    }
  };

  return (
    <DrawerBodyComponent
      scrolled={scrolled}
      as={component}
      theme={theme}
      ref={ref}
      {...otherProps}>
      <DrawerBodyScrollComponent
        onScroll={handleScrolled}
        theme={theme}>
        {children}
      </DrawerBodyScrollComponent>
    </DrawerBodyComponent>
  );
});

DrawerBody.displayName = "DrawerBody";

export default DrawerBody;
