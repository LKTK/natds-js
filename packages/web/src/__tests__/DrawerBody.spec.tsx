import React from 'react';
import { expect } from 'chai';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import DrawerBody, { DrawerBodyComponent, DrawerBodyScrollComponent } from '../Components/DrawerBody';

describe('DrawerBody component', () => {
  test('should match snapshot - DrawerBody children', () => {
    const children = (<div>some custom children</div>);
    const component = renderer.create(<DrawerBody children={children} />);

    expect(component).matchSnapshot('DrawerBody with children snapshot');
  });

  test('should match snapshot - DrawerBody custom component', () => {
    const children = (<div>some custom children</div>);
    const component = renderer.create(<DrawerBody children={children} component="section" />);

    expect(component).matchSnapshot('DrawerBody with custom component snapshot');
  });

  test('should custom children prevail over the properties', () => {
    const children = (<div className="custom">some custom children</div>);
    const component = mount(<DrawerBody children={children} />);
    const event: any = { target: { scrollTop: 10 } };

    component.find(DrawerBodyScrollComponent).simulate('scroll', event);

    expect(component.find(DrawerBodyComponent).prop('scrolled')).to.equal(true);
  });
});
