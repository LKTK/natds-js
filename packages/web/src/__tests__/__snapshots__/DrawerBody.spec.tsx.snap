// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrawerBody component should match snapshot - DrawerBody children: DrawerBody with children snapshot 1`] = `
<div
  className="sc-bdVaJa jTZBWV"
>
  <div
    className="sc-bwzfXH jrYved"
    onScroll={[Function]}
  >
    <div>
      some custom children
    </div>
  </div>
</div>
`;

exports[`DrawerBody component should match snapshot - DrawerBody custom component: DrawerBody with custom component snapshot 1`] = `
<section
  className="sc-bdVaJa jTZBWV"
>
  <div
    className="sc-bwzfXH jrYved"
    onScroll={[Function]}
  >
    <div>
      some custom children
    </div>
  </div>
</section>
`;
